File: src/IdentityBraces/Core/TraitCatalog.cs
using System.Collections.Generic;

namespace IdentityBraces.Core
{
    /// <summary>One trait's metadata: what it is called, which layer it sits on, how often.</summary>
    internal sealed class TraitInfo
    {
        public string Id;
        public TraitLayer Layer;
        public string Name;
        public int DefaultPercent;
        public string Description;
    }

    /// <summary>
    /// Every trait the extension knows about.
    /// </summary>
    /// <remarks>
    /// The single list driving the roll, the settings file, the presets and the options UI.
    /// Adding a trait is one row here plus one draw function in the drawing registry, and
    /// nothing else changes — which is the entire reason the registry exists.
    /// <para>
    /// Most default to zero. A fresh install should look like the extension people were
    /// shown rather than every switch at once; the presets are there to turn the rest on.
    /// </para>
    /// </remarks>
    internal static class TraitCatalog
    {
        public static readonly IList<TraitInfo> All = new List<TraitInfo>
        {
            new TraitInfo { Id = TraitIds.Question, Layer = TraitLayer.Body, Name = "Question mark", DefaultPercent = 6, Description = "Renders ? instead of the brace." },
            new TraitInfo { Id = TraitIds.UnicodeVariant, Layer = TraitLayer.Body, Name = "Unicode variant", DefaultPercent = 0, Description = "Picks an exotic bracket from the same family." },
            new TraitInfo { Id = TraitIds.WrongBracket, Layer = TraitLayer.Body, Name = "Wrong bracket", DefaultPercent = 0, Description = "Draws a different bracket type entirely." },
            new TraitInfo { Id = TraitIds.Mirrored, Layer = TraitLayer.Body, Name = "Mirrored", DefaultPercent = 0, Description = "Flipped horizontally, so openers close." },
            new TraitInfo { Id = TraitIds.UpsideDown, Layer = TraitLayer.Body, Name = "Upside down", DefaultPercent = 0, Description = "Rotated a half turn in place." },
            new TraitInfo { Id = TraitIds.Emoji, Layer = TraitLayer.Body, Name = "Emoji", DefaultPercent = 0, Description = "Substitutes an emoji. Ignores the palette." },
            new TraitInfo { Id = TraitIds.ForeignFont, Layer = TraitLayer.Body, Name = "Foreign font", DefaultPercent = 0, Description = "Drawn in Comic Sans while everything else stays monospaced." },
            new TraitInfo { Id = TraitIds.Subscript, Layer = TraitLayer.Body, Name = "Subscript", DefaultPercent = 0, Description = "Small, and dropped below the baseline." },
            new TraitInfo { Id = TraitIds.Catgirl, Layer = TraitLayer.Creature, Name = "Cat", DefaultPercent = 4, Description = "Two ears above the glyph." },
            new TraitInfo { Id = TraitIds.Bunny, Layer = TraitLayer.Creature, Name = "Bunny", DefaultPercent = 0, Description = "Tall narrow ears, one flopped at the tip." },
            new TraitInfo { Id = TraitIds.Devil, Layer = TraitLayer.Creature, Name = "Devil", DefaultPercent = 0, Description = "Curved horns and a barbed tail." },
            new TraitInfo { Id = TraitIds.Angel, Layer = TraitLayer.Creature, Name = "Angel", DefaultPercent = 0, Description = "A floating halo ring." },
            new TraitInfo { Id = TraitIds.Ghost, Layer = TraitLayer.Creature, Name = "Ghost", DefaultPercent = 0, Description = "Scalloped lower edge, drawn faint." },
            new TraitInfo { Id = TraitIds.Wizard, Layer = TraitLayer.Creature, Name = "Wizard", DefaultPercent = 0, Description = "A pointed hat with a brim and a star." },
            new TraitInfo { Id = TraitIds.Vampire, Layer = TraitLayer.Creature, Name = "Vampire", DefaultPercent = 0, Description = "Two fangs below the glyph." },
            new TraitInfo { Id = TraitIds.Bee, Layer = TraitLayer.Creature, Name = "Bee", DefaultPercent = 0, Description = "Banded body with a pair of wings." },
            new TraitInfo { Id = TraitIds.Frog, Layer = TraitLayer.Creature, Name = "Frog", DefaultPercent = 0, Description = "Two bulging eyes on the top curve." },
            new TraitInfo { Id = TraitIds.Snake, Layer = TraitLayer.Creature, Name = "Snake", DefaultPercent = 0, Description = "A forked tongue flicking sideways." },
            new TraitInfo { Id = TraitIds.Crab, Layer = TraitLayer.Creature, Name = "Crab", DefaultPercent = 0, Description = "A pincer on each flank." },
            new TraitInfo { Id = TraitIds.Bat, Layer = TraitLayer.Creature, Name = "Bat", DefaultPercent = 0, Description = "Scalloped wings either side." },
            new TraitInfo { Id = TraitIds.Spider, Layer = TraitLayer.Creature, Name = "Spider", DefaultPercent = 0, Description = "Legs radiating outward." },
            new TraitInfo { Id = TraitIds.Fox, Layer = TraitLayer.Creature, Name = "Fox", DefaultPercent = 0, Description = "Sharp ears and a heavy tail." },
            new TraitInfo { Id = TraitIds.Wolf, Layer = TraitLayer.Creature, Name = "Wolf", DefaultPercent = 0, Description = "Outswept ears and a snout." },
            new TraitInfo { Id = TraitIds.Dragon, Layer = TraitLayer.Creature, Name = "Dragon", DefaultPercent = 0, Description = "Swept horns and a folded wing." },
            new TraitInfo { Id = TraitIds.Unicorn, Layer = TraitLayer.Creature, Name = "Unicorn", DefaultPercent = 0, Description = "A single horn, dead centre." },
            new TraitInfo { Id = TraitIds.Penguin, Layer = TraitLayer.Creature, Name = "Penguin", DefaultPercent = 0, Description = "A beak and two flippers." },
            new TraitInfo { Id = TraitIds.Owl, Layer = TraitLayer.Creature, Name = "Owl", DefaultPercent = 0, Description = "Two oversized concentric eyes." },
            new TraitInfo { Id = TraitIds.Cthulhu, Layer = TraitLayer.Creature, Name = "Cthulhu", DefaultPercent = 0, Description = "Tentacles drooping and slowly waving." },
            new TraitInfo { Id = TraitIds.Slime, Layer = TraitLayer.Creature, Name = "Slime", DefaultPercent = 0, Description = "A drip that forms, falls and resets." },
            new TraitInfo { Id = TraitIds.Mushroom, Layer = TraitLayer.Creature, Name = "Mushroom", DefaultPercent = 0, Description = "A spotted cap sitting on top." },
            new TraitInfo { Id = TraitIds.Cactus, Layer = TraitLayer.Creature, Name = "Cactus", DefaultPercent = 0, Description = "Spines along both flanks and a flower." },
            new TraitInfo { Id = TraitIds.Robot, Layer = TraitLayer.Creature, Name = "Robot", DefaultPercent = 0, Description = "An antenna with a blinking bulb." },
            new TraitInfo { Id = TraitIds.Pirate, Layer = TraitLayer.Creature, Name = "Pirate", DefaultPercent = 0, Description = "An eyepatch band and a bandana knot." },
            new TraitInfo { Id = TraitIds.ThighHighs, Layer = TraitLayer.Costume, Name = "Thigh highs", DefaultPercent = 4, Description = "The lower stroke recoloured, with a welt stripe." },
            new TraitInfo { Id = TraitIds.TopHat, Layer = TraitLayer.Costume, Name = "Top hat", DefaultPercent = 0, Description = "A brim and a block above it." },
            new TraitInfo { Id = TraitIds.Crown, Layer = TraitLayer.Costume, Name = "Crown", DefaultPercent = 0, Description = "A three-point zigzag." },
            new TraitInfo { Id = TraitIds.Sunglasses, Layer = TraitLayer.Costume, Name = "Sunglasses", DefaultPercent = 0, Description = "A dark bar with a bridge notch." },
            new TraitInfo { Id = TraitIds.Scarf, Layer = TraitLayer.Costume, Name = "Scarf", DefaultPercent = 0, Description = "A band with one tail streaming out." },
            new TraitInfo { Id = TraitIds.Bowtie, Layer = TraitLayer.Costume, Name = "Bow tie", DefaultPercent = 0, Description = "Two triangles meeting at the waist." },
            new TraitInfo { Id = TraitIds.Cape, Layer = TraitLayer.Costume, Name = "Cape", DefaultPercent = 0, Description = "A shape behind the stroke." },
            new TraitInfo { Id = TraitIds.PartyHat, Layer = TraitLayer.Costume, Name = "Party hat", DefaultPercent = 0, Description = "A striped cone with a pom." },
            new TraitInfo { Id = TraitIds.Headphones, Layer = TraitLayer.Costume, Name = "Headphones", DefaultPercent = 0, Description = "An arc over the top with two pads." },
            new TraitInfo { Id = TraitIds.FlowerCrown, Layer = TraitLayer.Costume, Name = "Flower crown", DefaultPercent = 0, Description = "Three dots of differing hues." },
            new TraitInfo { Id = TraitIds.Beanie, Layer = TraitLayer.Costume, Name = "Beanie", DefaultPercent = 0, Description = "A rounded cap with a rolled brim." },
            new TraitInfo { Id = TraitIds.Monocle, Layer = TraitLayer.Costume, Name = "Monocle", DefaultPercent = 0, Description = "A ring with a chain hanging down." },
            new TraitInfo { Id = TraitIds.Necktie, Layer = TraitLayer.Costume, Name = "Necktie", DefaultPercent = 0, Description = "A knot and a taper down the front." },
            new TraitInfo { Id = TraitIds.Backpack, Layer = TraitLayer.Costume, Name = "Backpack", DefaultPercent = 0, Description = "A rounded box behind, with straps." },
            new TraitInfo { Id = TraitIds.Wings, Layer = TraitLayer.Costume, Name = "Wings", DefaultPercent = 0, Description = "Feathered arcs on both flanks." },
            new TraitInfo { Id = TraitIds.Armour, Layer = TraitLayer.Costume, Name = "Armour", DefaultPercent = 0, Description = "A plated band with a rivet." },
            new TraitInfo { Id = TraitIds.Bandage, Layer = TraitLayer.Costume, Name = "Bandage", DefaultPercent = 0, Description = "A crossed plaster over the stroke." },
            new TraitInfo { Id = TraitIds.Moustache, Layer = TraitLayer.Costume, Name = "Moustache", DefaultPercent = 0, Description = "A curled bar across the middle." },
            new TraitInfo { Id = TraitIds.ColourCycle, Layer = TraitLayer.Motion, Name = "Colour cycle", DefaultPercent = 8, Description = "Travels the hue wheel forever." },
            new TraitInfo { Id = TraitIds.Wobble, Layer = TraitLayer.Motion, Name = "Wobble", DefaultPercent = 0, Description = "Rotates a few degrees either side." },
            new TraitInfo { Id = TraitIds.Bounce, Layer = TraitLayer.Motion, Name = "Bounce", DefaultPercent = 0, Description = "A short vertical hop on a loop." },
            new TraitInfo { Id = TraitIds.Breathe, Layer = TraitLayer.Motion, Name = "Breathe", DefaultPercent = 0, Description = "Scales slowly between 95 and 105 percent." },
            new TraitInfo { Id = TraitIds.Heartbeat, Layer = TraitLayer.Motion, Name = "Heartbeat", DefaultPercent = 0, Description = "Two quick pulses, then a rest." },
            new TraitInfo { Id = TraitIds.Shiver, Layer = TraitLayer.Motion, Name = "Shiver", DefaultPercent = 0, Description = "Sub-pixel jitter on both axes." },
            new TraitInfo { Id = TraitIds.Blink, Layer = TraitLayer.Motion, Name = "Blink", DefaultPercent = 0, Description = "Fades out and back at irregular intervals." },
            new TraitInfo { Id = TraitIds.Spin, Layer = TraitLayer.Motion, Name = "Spin", DefaultPercent = 0, Description = "Continuous slow rotation." },
            new TraitInfo { Id = TraitIds.Flip, Layer = TraitLayer.Motion, Name = "Flip", DefaultPercent = 0, Description = "Snaps a half turn every few seconds." },
            new TraitInfo { Id = TraitIds.Glitch, Layer = TraitLayer.Motion, Name = "Glitch", DefaultPercent = 0, Description = "Random small offsets with colour fringing." },
            new TraitInfo { Id = TraitIds.Sparkle, Layer = TraitLayer.Motion, Name = "Sparkle", DefaultPercent = 0, Description = "Four-point stars appearing and fading." },
            new TraitInfo { Id = TraitIds.Drift, Layer = TraitLayer.Motion, Name = "Drift", DefaultPercent = 0, Description = "Wanders off its cell and slowly returns." },
            new TraitInfo { Id = TraitIds.Shimmer, Layer = TraitLayer.Motion, Name = "Shimmer", DefaultPercent = 0, Description = "A bright band sweeping down the stroke." },
            new TraitInfo { Id = TraitIds.Flicker, Layer = TraitLayer.Motion, Name = "Flicker", DefaultPercent = 0, Description = "Firelight brightness noise." },
            new TraitInfo { Id = TraitIds.Wave, Layer = TraitLayer.Motion, Name = "Wave", DefaultPercent = 0, Description = "Neighbours bounce in sequence along the line." },
            new TraitInfo { Id = TraitIds.Typewriter, Layer = TraitLayer.Motion, Name = "Typewriter", DefaultPercent = 0, Description = "Fades in once on first appearance." },
            new TraitInfo { Id = TraitIds.Fire, Layer = TraitLayer.Effect, Name = "On fire", DefaultPercent = 0, Description = "Flames licking up from the glyph." },
            new TraitInfo { Id = TraitIds.GradientFill, Layer = TraitLayer.Effect, Name = "Gradient fill", DefaultPercent = 0, Description = "A vertical ramp through the stroke." },
            new TraitInfo { Id = TraitIds.BoldItalic, Layer = TraitLayer.Effect, Name = "Bold or italic", DefaultPercent = 0, Description = "Per-brace weight and slant." },
            new TraitInfo { Id = TraitIds.Tilted, Layer = TraitLayer.Effect, Name = "Tilted", DefaultPercent = 0, Description = "A fixed random lean, stable per brace." },
            new TraitInfo { Id = TraitIds.Underline, Layer = TraitLayer.Effect, Name = "Underline", DefaultPercent = 0, Description = "A squiggle beneath, like a spelling error." },
            new TraitInfo { Id = TraitIds.Shadow, Layer = TraitLayer.Effect, Name = "Shadow", DefaultPercent = 0, Description = "A hard offset duplicate behind the glyph." },
            new TraitInfo { Id = TraitIds.Distressed, Layer = TraitLayer.Effect, Name = "Distressed", DefaultPercent = 0, Description = "Sweat beads and an unsteady shake. Also forced on by the complexity warning." },
            new TraitInfo { Id = TraitIds.FleeCursor, Layer = TraitLayer.Effect, Name = "Flee the cursor", DefaultPercent = 0, Description = "Edges away as the caret nears." },
            new TraitInfo { Id = TraitIds.Named, Layer = TraitLayer.Effect, Name = "Named", DefaultPercent = 0, Description = "A stable generated name, shown on hover." },
            new TraitInfo { Id = TraitIds.Nocturnal, Layer = TraitLayer.Effect, Name = "Nocturnal", DefaultPercent = 0, Description = "Droops as the session wears on." },
            new TraitInfo { Id = TraitIds.Seasonal, Layer = TraitLayer.Effect, Name = "Seasonal", DefaultPercent = 0, Description = "Dresses for the time of year." },
            new TraitInfo { Id = TraitIds.BuildReactive, Layer = TraitLayer.Effect, Name = "Build reactive", DefaultPercent = 0, Description = "Celebrates a green build, sulks at a red one." },
            new TraitInfo { Id = TraitIds.SwapPlaces, Layer = TraitLayer.Effect, Name = "Swap places", DefaultPercent = 0, Description = "Two braces on a line trade positions, briefly." },
            new TraitInfo { Id = TraitIds.Drunk, Layer = TraitLayer.Effect, Name = "Drunk", DefaultPercent = 0, Description = "The lean grows over the session, then resets." },
            new TraitInfo { Id = TraitIds.Gravity, Layer = TraitLayer.Effect, Name = "Gravity", DefaultPercent = 0, Description = "Sags toward the bottom of its cell over time." },
            new TraitInfo { Id = TraitIds.StageFright, Layer = TraitLayer.Effect, Name = "Stage fright", DefaultPercent = 0, Description = "Hides while the caret is on its line." },
            new TraitInfo { Id = TraitIds.Mitosis, Layer = TraitLayer.Effect, Name = "Mitosis", DefaultPercent = 0, Description = "Rarely splits in two; one half drifts away." },
            new TraitInfo { Id = TraitIds.TableFlip, Layer = TraitLayer.Effect, Name = "Table flip", DefaultPercent = 0, Description = "Flips a table into adjacent whitespace, then tidies up." },
            new TraitInfo { Id = TraitIds.FireBrigade, Layer = TraitLayer.Effect, Name = "Fire brigade", DefaultPercent = 0, Description = "Braces crew a fire truck and put out a burning brace." },
        };

        public static TraitInfo Find(string id)
        {
            for (int i = 0; i < All.Count; i++)
            {
                if (All[i].Id == id)
                {
                    return All[i];
                }
            }

            return null;
        }
    }
}

File: jetbrains/src/main/kotlin/identitybraces/core/TraitCatalog.kt
package identitybraces.core

/** One trait's metadata: what it is called, which layer it sits on, how often. */
class TraitInfo(
    val id: String,
    val layer: TraitLayer,
    val name: String,
    val defaultPercent: Int,
    val description: String,
)

/** One trait's identity and how often it should come up. */
class TraitWeight(val id: String, val layer: TraitLayer, val percent: Int)

/**
 * Every trait the extension knows about, in the order the roll walks them.
 *
 * The single list driving the roll, the settings, the presets and the settings UI. Adding a
 * trait is one row here plus one draw function in the drawing registry, and nothing else
 * changes. The *order* is part of the identity contract — the flattened trait table walks it
 * in sequence, so a reshuffle would hand every brace a different personality — which is why
 * this is a list and not a map.
 */
object TraitCatalog {
    val all: List<TraitInfo> = listOf(
        TraitInfo(TraitIds.QUESTION, TraitLayer.Body, "Question mark", 6, "Renders ? instead of the brace."),
        TraitInfo(TraitIds.UNICODE_VARIANT, TraitLayer.Body, "Unicode variant", 0, "Picks an exotic bracket from the same family."),
        TraitInfo(TraitIds.WRONG_BRACKET, TraitLayer.Body, "Wrong bracket", 0, "Draws a different bracket type entirely."),
        TraitInfo(TraitIds.MIRRORED, TraitLayer.Body, "Mirrored", 0, "Flipped horizontally, so openers close."),
        TraitInfo(TraitIds.UPSIDE_DOWN, TraitLayer.Body, "Upside down", 0, "Rotated a half turn in place."),
        TraitInfo(TraitIds.EMOJI, TraitLayer.Body, "Emoji", 0, "Substitutes an emoji. Ignores the palette."),
        TraitInfo(TraitIds.FOREIGN_FONT, TraitLayer.Body, "Foreign font", 0, "Drawn in Comic Sans while everything else stays monospaced."),
        TraitInfo(TraitIds.SUBSCRIPT, TraitLayer.Body, "Subscript", 0, "Small, and dropped below the baseline."),
        TraitInfo(TraitIds.CATGIRL, TraitLayer.Creature, "Cat", 4, "Two ears above the glyph."),
        TraitInfo(TraitIds.BUNNY, TraitLayer.Creature, "Bunny", 0, "Tall narrow ears, one flopped at the tip."),
        TraitInfo(TraitIds.DEVIL, TraitLayer.Creature, "Devil", 0, "Curved horns and a barbed tail."),
        TraitInfo(TraitIds.ANGEL, TraitLayer.Creature, "Angel", 0, "A floating halo ring."),
        TraitInfo(TraitIds.GHOST, TraitLayer.Creature, "Ghost", 0, "Scalloped lower edge, drawn faint."),
        TraitInfo(TraitIds.WIZARD, TraitLayer.Creature, "Wizard", 0, "A pointed hat with a brim and a star."),
        TraitInfo(TraitIds.VAMPIRE, TraitLayer.Creature, "Vampire", 0, "Two fangs below the glyph."),
        TraitInfo(TraitIds.BEE, TraitLayer.Creature, "Bee", 0, "Banded body with a pair of wings."),
        TraitInfo(TraitIds.FROG, TraitLayer.Creature, "Frog", 0, "Two bulging eyes on the top curve."),
        TraitInfo(TraitIds.SNAKE, TraitLayer.Creature, "Snake", 0, "A forked tongue flicking sideways."),
        TraitInfo(TraitIds.CRAB, TraitLayer.Creature, "Crab", 0, "A pincer on each flank."),
        TraitInfo(TraitIds.BAT, TraitLayer.Creature, "Bat", 0, "Scalloped wings either side."),
        TraitInfo(TraitIds.SPIDER, TraitLayer.Creature, "Spider", 0, "Legs radiating outward."),
        TraitInfo(TraitIds.FOX, TraitLayer.Creature, "Fox", 0, "Sharp ears and a heavy tail."),
        TraitInfo(TraitIds.WOLF, TraitLayer.Creature, "Wolf", 0, "Outswept ears and a snout."),
        TraitInfo(TraitIds.DRAGON, TraitLayer.Creature, "Dragon", 0, "Swept horns and a folded wing."),
        TraitInfo(TraitIds.UNICORN, TraitLayer.Creature, "Unicorn", 0, "A single horn, dead centre."),
        TraitInfo(TraitIds.PENGUIN, TraitLayer.Creature, "Penguin", 0, "A beak and two flippers."),
        TraitInfo(TraitIds.OWL, TraitLayer.Creature, "Owl", 0, "Two oversized concentric eyes."),
        TraitInfo(TraitIds.CTHULHU, TraitLayer.Creature, "Cthulhu", 0, "Tentacles drooping and slowly waving."),
        TraitInfo(TraitIds.SLIME, TraitLayer.Creature, "Slime", 0, "A drip that forms, falls and resets."),
        TraitInfo(TraitIds.MUSHROOM, TraitLayer.Creature, "Mushroom", 0, "A spotted cap sitting on top."),
        TraitInfo(TraitIds.CACTUS, TraitLayer.Creature, "Cactus", 0, "Spines along both flanks and a flower."),
        TraitInfo(TraitIds.ROBOT, TraitLayer.Creature, "Robot", 0, "An antenna with a blinking bulb."),
        TraitInfo(TraitIds.PIRATE, TraitLayer.Creature, "Pirate", 0, "An eyepatch band and a bandana knot."),
        TraitInfo(TraitIds.THIGH_HIGHS, TraitLayer.Costume, "Thigh highs", 4, "The lower stroke recoloured, with a welt stripe."),
        TraitInfo(TraitIds.TOP_HAT, TraitLayer.Costume, "Top hat", 0, "A brim and a block above it."),
        TraitInfo(TraitIds.CROWN, TraitLayer.Costume, "Crown", 0, "A three-point zigzag."),
        TraitInfo(TraitIds.SUNGLASSES, TraitLayer.Costume, "Sunglasses", 0, "A dark bar with a bridge notch."),
        TraitInfo(TraitIds.SCARF, TraitLayer.Costume, "Scarf", 0, "A band with one tail streaming out."),
        TraitInfo(TraitIds.BOWTIE, TraitLayer.Costume, "Bow tie", 0, "Two triangles meeting at the waist."),
        TraitInfo(TraitIds.CAPE, TraitLayer.Costume, "Cape", 0, "A shape behind the stroke."),
        TraitInfo(TraitIds.PARTY_HAT, TraitLayer.Costume, "Party hat", 0, "A striped cone with a pom."),
        TraitInfo(TraitIds.HEADPHONES, TraitLayer.Costume, "Headphones", 0, "An arc over the top with two pads."),
        TraitInfo(TraitIds.FLOWER_CROWN, TraitLayer.Costume, "Flower crown", 0, "Three dots of differing hues."),
        TraitInfo(TraitIds.BEANIE, TraitLayer.Costume, "Beanie", 0, "A rounded cap with a rolled brim."),
        TraitInfo(TraitIds.MONOCLE, TraitLayer.Costume, "Monocle", 0, "A ring with a chain hanging down."),
        TraitInfo(TraitIds.NECKTIE, TraitLayer.Costume, "Necktie", 0, "A knot and a taper down the front."),
        TraitInfo(TraitIds.BACKPACK, TraitLayer.Costume, "Backpack", 0, "A rounded box behind, with straps."),
        TraitInfo(TraitIds.WINGS, TraitLayer.Costume, "Wings", 0, "Feathered arcs on both flanks."),
        TraitInfo(TraitIds.ARMOUR, TraitLayer.Costume, "Armour", 0, "A plated band with a rivet."),
        TraitInfo(TraitIds.BANDAGE, TraitLayer.Costume, "Bandage", 0, "A crossed plaster over the stroke."),
        TraitInfo(TraitIds.MOUSTACHE, TraitLayer.Costume, "Moustache", 0, "A curled bar across the middle."),
        TraitInfo(TraitIds.COLOUR_CYCLE, TraitLayer.Motion, "Colour cycle", 8, "Travels the hue wheel forever."),
        TraitInfo(TraitIds.WOBBLE, TraitLayer.Motion, "Wobble", 0, "Rotates a few degrees either side."),
        TraitInfo(TraitIds.BOUNCE, TraitLayer.Motion, "Bounce", 0, "A short vertical hop on a loop."),
        TraitInfo(TraitIds.BREATHE, TraitLayer.Motion, "Breathe", 0, "Scales slowly between 95 and 105 percent."),
        TraitInfo(TraitIds.HEARTBEAT, TraitLayer.Motion, "Heartbeat", 0, "Two quick pulses, then a rest."),
        TraitInfo(TraitIds.SHIVER, TraitLayer.Motion, "Shiver", 0, "Sub-pixel jitter on both axes."),
        TraitInfo(TraitIds.BLINK, TraitLayer.Motion, "Blink", 0, "Fades out and back at irregular intervals."),
        TraitInfo(TraitIds.SPIN, TraitLayer.Motion, "Spin", 0, "Continuous slow rotation."),
        TraitInfo(TraitIds.FLIP, TraitLayer.Motion, "Flip", 0, "Snaps a half turn every few seconds."),
        TraitInfo(TraitIds.GLITCH, TraitLayer.Motion, "Glitch", 0, "Random small offsets with colour fringing."),
        TraitInfo(TraitIds.SPARKLE, TraitLayer.Motion, "Sparkle", 0, "Four-point stars appearing and fading."),
        TraitInfo(TraitIds.DRIFT, TraitLayer.Motion, "Drift", 0, "Wanders off its cell and slowly returns."),
        TraitInfo(TraitIds.SHIMMER, TraitLayer.Motion, "Shimmer", 0, "A bright band sweeping down the stroke."),
        TraitInfo(TraitIds.FLICKER, TraitLayer.Motion, "Flicker", 0, "Firelight brightness noise."),
        TraitInfo(TraitIds.WAVE, TraitLayer.Motion, "Wave", 0, "Neighbours bounce in sequence along the line."),
        TraitInfo(TraitIds.TYPEWRITER, TraitLayer.Motion, "Typewriter", 0, "Fades in once on first appearance."),
        TraitInfo(TraitIds.FIRE, TraitLayer.Effect, "On fire", 0, "Flames licking up from the glyph."),
        TraitInfo(TraitIds.GRADIENT_FILL, TraitLayer.Effect, "Gradient fill", 0, "A vertical ramp through the stroke."),
        TraitInfo(TraitIds.BOLD_ITALIC, TraitLayer.Effect, "Bold or italic", 0, "Per-brace weight and slant."),
        TraitInfo(TraitIds.TILTED, TraitLayer.Effect, "Tilted", 0, "A fixed random lean, stable per brace."),
        TraitInfo(TraitIds.UNDERLINE, TraitLayer.Effect, "Underline", 0, "A squiggle beneath, like a spelling error."),
        TraitInfo(TraitIds.SHADOW, TraitLayer.Effect, "Shadow", 0, "A hard offset duplicate behind the glyph."),
        TraitInfo(TraitIds.DISTRESSED, TraitLayer.Effect, "Distressed", 0, "Sweat beads and an unsteady shake. Also forced on by the complexity warning."),
        TraitInfo(TraitIds.FLEE_CURSOR, TraitLayer.Effect, "Flee the cursor", 0, "Edges away as the caret nears."),
        TraitInfo(TraitIds.NAMED, TraitLayer.Effect, "Named", 0, "A stable generated name, shown on hover."),
        TraitInfo(TraitIds.NOCTURNAL, TraitLayer.Effect, "Nocturnal", 0, "Droops as the session wears on."),
        TraitInfo(TraitIds.SEASONAL, TraitLayer.Effect, "Seasonal", 0, "Dresses for the time of year."),
        TraitInfo(TraitIds.BUILD_REACTIVE, TraitLayer.Effect, "Build reactive", 0, "Celebrates a green build, sulks at a red one."),
        TraitInfo(TraitIds.SWAP_PLACES, TraitLayer.Effect, "Swap places", 0, "Two braces on a line trade positions, briefly."),
        TraitInfo(TraitIds.DRUNK, TraitLayer.Effect, "Drunk", 0, "The lean grows over the session, then resets."),
        TraitInfo(TraitIds.GRAVITY, TraitLayer.Effect, "Gravity", 0, "Sags toward the bottom of its cell over time."),
        TraitInfo(TraitIds.STAGE_FRIGHT, TraitLayer.Effect, "Stage fright", 0, "Hides while the caret is on its line."),
        TraitInfo(TraitIds.MITOSIS, TraitLayer.Effect, "Mitosis", 0, "Rarely splits in two; one half drifts away."),
        TraitInfo(TraitIds.TABLE_FLIP, TraitLayer.Effect, "Table flip", 0, "Flips a table into adjacent whitespace, then tidies up."),
        TraitInfo(TraitIds.FIRE_BRIGADE, TraitLayer.Effect, "Fire brigade", 0, "Braces crew a fire truck and put out a burning brace."),
    )

    private val byId: Map<String, TraitInfo> = all.associateBy { it.id }

    fun find(id: String): TraitInfo? = byId[id]

    /** The catalogue's own defaults, for tests and for a first run. */
    fun defaultWeights(): List<TraitWeight> = all.map { TraitWeight(it.id, it.layer, it.defaultPercent) }

    /**
     * Catalogue defaults with [overrides] applied by id, clamped to 0–100.
     *
     * Kept in catalogue order regardless of the order the overrides arrive in — see the note
     * on [all].
     */
    fun weightsFrom(overrides: Map<String, Int>?): List<TraitWeight> = all.map { info ->
        val percent = overrides?.get(info.id) ?: info.defaultPercent
        TraitWeight(info.id, info.layer, percent.coerceIn(0, 100))
    }
}
